userspace/libsinsp/metrics_collector.h

```diff
@@ -0,0 +1,213 @@
+	*/
+	void snapshot();
+
+	/*!
+	\brief Method to get a const reference to m_metrics vector
+	*/
+	const std::vector<metrics_v2>& get_metrics() const;
+
+	/*!
+	\brief Method to convert a metric to the text-based Prometheus format.
+	 * Reference: https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md
+	 * Note: The design idea is to expose Prometheus metrics by piping text-based formats to new line-delimited fields 
+	 * exposed at /metrics in Falco or a similar approach, eliminating the need for implementing a full Prometheus client.
+	 * Example:
+	 * test.memory_used_host{raw_name="memory_used_host", unit="MEMORY_MEGABYTES", metric_type="NON_MONOTONIC_CURRENT"} 15096.000000 1706490801547502000
+	 * 
+	 * This method is a work in progress.
+	*/
+	const std::string convert_metric_to_prometheus_text(std::string metric_name, metrics_v2 metric) const;
+
+	/*!
+	\brief Method to convert memory units; tied to metrics_v2 definitions
+	*/
+	template <typename T>
+	double convert_memory(metrics_v2_value_unit source_unit, metrics_v2_value_unit dest_unit, T val)
+	{
+		double factor = 1;
+		switch(source_unit)
+		{
+		case METRIC_VALUE_UNIT_MEMORY_BYTES:
+			factor = 1;
+			break;
+		case METRIC_VALUE_UNIT_MEMORY_KILOBYTES:
+			factor = 1024.;
+			break;
+		case METRIC_VALUE_UNIT_MEMORY_MEGABYTES:
+			factor = 1024. * 1024.;
+			break;
+		default:
+			return 0;
+		}
+
+		double bytes_val = val * factor;
+		switch(dest_unit)
+		{
+		case METRIC_VALUE_UNIT_MEMORY_BYTES:
+			return bytes_val;
+		case METRIC_VALUE_UNIT_MEMORY_KILOBYTES:
+			return std::round((bytes_val / 1024.) * 10.) / 10.; // round to 1 decimal
+		case METRIC_VALUE_UNIT_MEMORY_MEGABYTES:
+			return std::round((bytes_val / 1024. / 1024.) * 10.) / 10.; // round to 1 decimal
+		default:
+			return 0;
+		}
+		return 0;
+	}
+
+private:
+	sinsp* m_inspector;
+	uint32_t m_metrics_flags = METRICS_V2_KERNEL_COUNTERS | METRICS_V2_LIBBPF_STATS | METRICS_V2_RESOURCE_UTILIZATION | METRICS_V2_STATE_COUNTERS;
+	bool m_convert_memory_to_mb = true;
+	std::vector<metrics_v2> m_metrics;
+
+	void get_rss_vsz_pss_total_memory_and_open_fds(uint32_t &rss, uint32_t &vsz, uint32_t &pss, uint64_t &memory_used_host, uint64_t &open_fds_host);
+	void get_cpu_usage_and_total_procs(double start_time, double &cpu_usage_perc, double &cpu_usage_perc_total_host, uint32_t &procs_running_host);
+	uint64_t get_container_memory_usage() const;
+
+	template <typename T>
+	const metrics_v2 new_metric(const char* name, uint32_t flags, metrics_v2_value_type type, metrics_v2_value_unit unit, metrics_v2_metric_type metric_type, T val)
+	{
+		metrics_v2 metric;
+		strlcpy(metric.name, name, METRIC_NAME_MAX);
+		metric.flags = flags;
+		metric.type = type;
+		metric.unit = unit;
+		metric.metric_type = metric_type;
+		set_new_metric(metric, type, val);
+		return metric;
+	}
+
+	template <typename T>
+	void set_new_metric(metrics_v2& metric, metrics_v2_value_type type, T val)
```

Comment:
This also could be static, a free function or part of the logic of the constructor of `metrics_v2`.